### DISCUSSION TITLE
Rate limits on docker related command

### DISCUSSION THREAD

**Discussion by hmzdot** [reach-sh/reach-lang#1030] [Q&A] [answered]
Using `./reach update` and `./reach info` occasionally errors with:
```
./reach update
Exhausted rate-limited retries for:
 * reachsh/reach        after 7 attempts.
 * reachsh/reach-cli    after 7 attempts.
 * reachsh/react-runner after 7 attempts.
 * reachsh/rpc-server   after 7 attempts.
 * reachsh/runner       after 7 attempts.
 * reachsh/devnet-algo  after 7 attempts.
 * library/postgres     after 7 attempts.
 * reachsh/devnet-cfx   after 7 attempts.
 * reachsh/devnet-eth   after 7 attempts.
```
Note that I tried this in WSL2 but I remember getting the same problem on a Ubuntu 20.04 machine last week

**Comment by MetaB0y** [comment]
I have this problem sort of consistently e.g. I never can update but sometimes I have fewer lines e.g.
```
Exhausted rate-limited retries for:
 * reachsh/runner      after 7 attempts.
 * reachsh/devnet-algo after 7 attempts.
 * reachsh/devnet-cfx  after 7 attempts.
 ````

**Comment by kcelestinomaria** [comment]
Same problem today. I can't update my reach version. Trying to compile and run my whole reach program and I get this error
```
dApp/node_modules/@reach-sh/stdlib/dist/cjs/shared_impl.js:592
        throw Error("This Reach compiled ".concat(label, " does not match the expectations of this Reach standard library: expected ").concat(expected, ", but got ").concat(actual, "; ").concat(more));
              ^

Error: This Reach compiled JavaScript backend does not match the expectations of this Reach standard library: expected 16, but got 17; update your standard library and rerun!
```

But when I try to update i.e run ```./reach update```, this happens:

```
./reach update
Exhausted rate-limited retries for:
 * reachsh/reach        after 8 attempts.
 * reachsh/react-runner after 8 attempts.
 * reachsh/runner       after 8 attempts.
 * reachsh/devnet-algo  after 8 attempts.
 * reachsh/devnet-cfx   after 8 attempts.
```

**Comment by kcelestinomaria** [comment]
Could it be that ```./reach update``` updates the contract(```.rsh```) version and you are needed to update ```@reach-stdlib``` ? @zetsuboii @MetaB0y

**Comment by jeapostrophe** [comment]
@mattaudesse Ping

**Comment by mattaudesse** [comment]
For those of you who see this happen frequently it'd be helpful to know where in the world you're located (just country would be great) and whether you're behind some kind of shared business/government/school network.

I suspect NAT might account for some of these trouble reports.

Whatever the case, I'll consider better ways to manage DockerHub's rate-limiting.

**Reply by mattaudesse** [reply]
(You can DM me on Discord if you'd prefer not to share details publicly.)

**Reply by kcelestinomaria** [reply]
Hello, I was using gitpod to quickly run my setup @mattaudesse , might it be the reason? Cause gitpod is containerized

**Reply by mattaudesse** [reply]
Thanks @kcelestinomaria - it'd make sense if DockerHub throttled requests originating from a SaaS like gitpod since they're probably getting a lot of (non-Reach-related) traffic from them all the time.

Our best answer to bypass this limitation would be to simply ditch DockerHub and host our own image repository but that probably won't happen anytime soon.

If you can configure your own machine to run Reach instead of using gitpod I think you'll have more luck.

**Reply by kcelestinomaria** [reply]
Noted @mattaudesse . Many thanks for this, I'll work it out.

**Comment by DanBurton** [comment, marked as answer]
You can try this workaround:

```
docker pull reachsh/reach-cli:latest
./reach update
```
